text/0000-generalized-arity-tuples.md

```diff
@@ -0,0 +1,277 @@
+        if let (Some(elem), Some(tail)) = (self.elem.join(), self.tail.join()) {
+            Some(Tuple::new(elem, tail))
+        } else {
+            None
+        }
+    }
+}
+```
+
+Note that `Tuple<ELEM, TAIL>` is a desugared representation of the tuple `(ELEM, TAIL.0, ..., TAIL.n-1)` with arity `n + 1` where `TAIL` is a tuple with arity `n`.
+
+## Why are we already done here?
+
+Consider the type `(Option<i32>, (Option<bool>, Option<&str>))` from the requirements above. This type is just syntactic sugar for the type `Tuple<Option<i32>, Tuple<Tuple<Option<bool>, Tuple<Option<&str>, ()>>, ()>>`.
+
+Now, apply the provided trait implementations step by step on the desugared type. The resulting associated type `Join::Joined` evaluates to `Tuple<i32, Tuple<Tuple<bool, Tuple<&str, ()>>, ()>>`. But this is just the desugared version of `(i32, (bool, &str))`.
+
+This illustrates how the mechanims works on the type level.
+
+## More examples / Advanced type mappings
+
+Generalized implementations on `Tuple<HEAD, TAIL>` are not restricted to mappings between tuples of the same shape. The following example demonstrates how a `last` function can be realized that works with any tuple:
+
+```rust
+trait Last {
+    type Last;
+
+    fn last(self) -> Self::Last;
+}
+
+impl<ELEM> Last for (ELEM,) {
+    type Last = ELEM;
+
+    fn last(self) -> Self::Last {
+        self.elem
+    }
+}
+
+impl<ELEM, TAIL: Last> Last for Tuple<ELEM, TAIL> {
+    type Last = TAIL::Last;
+
+    fn last(self) -> Self::Last {
+        self.tail.last()
+    }
+}
+```
+
+Correcty, the compiler rejects empty tuples while tuples of other sizes are accepted:
+
+```rust
+().last();       // does not compile: no method named `last` found for type `()` in the current scope
+(1,).last();     // returns 1
+(1, "two").last(); // returns "two"
+etc.
+```
+
+The last example demonstrates how every second element of a tuple can be removed:
+
+```rust
+trait Halve {
+    type Output;
+
+    fn halve(self) -> Self::Output;
+}
+
+impl Halve for () {
+    type Output = ();
+
+    fn halve(self) {}
+}
+
+impl<ELEM1, ELEM2, TAIL: Halve> Halve for Tuple<ELEM1, Tuple<ELEM2, TAIL>> {
+    type Output = Tuple<ELEM1, TAIL>;
+
+    fn halve(self) -> Self::Output {
+        Tuple::new(self.elem, self.tail.tail)
+    }
+}
+```
+
+Results:
+
+```rust
+().halve();                   // returns ()
+(1,).halve()                  // does not compile: no method named `halve` found for type `(i32,)` in the current scope
+(1, "two").halve();           // returns (1,)
+(1, "two", 3.0, '4').halve(); // returns (1, 3.0)
+etc.
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The idea of this RFC to leave the tuple notation as is but to treat tuple type expressions as aliases according to the following pattern:
+
+```rust
+type () = (); // Not really an alias. Written down for completeness.
+type (A,) = Tuple<A, ()>;
+type (A, B) = Tuple<A, Tuple<B, ()>>;
```

Comment:
Unfortunately this desugaring is incompatible with the current unsizing rule.

1. We need allow `([u8],)`.
2. We also need to allow `(u8, [u8])`.

If `(A,)` is desugared as `Tuple<A, ()>`, this means `ELEM` must be relaxed as `ELEM: ?Sized`.

If `(A, B)` is desugared as `Tuple<A, (B,)>`, this means `TAIL` must be relaxed as `TAIL: ?Sized`.

But we cannot have two unsized fields in a structure (`struct Tuple<E: ?Sized, T: ?Sized> { head: E, tail: T }`.

Therefore, the tuple desugaring must terminate at `(A,)` and cannot be further desugared to `Tuple<A, ()>`. 

Alternatively, you could reverse the expansion direction, so that only the final field needs to be unsized.

```rust
struct Tuple<Init, Last: ?Sized> {
    init: Init,
    last: Last,
}

type () = ();
type (A,) = Tuple<(), A>;
type (A, B) = Tuple<Tuple<(), A>, B>;
type (A, B, C) = Tuple<Tuple<Tuple<(), A>, B>, C>;
```

Reply:
Good point! Easy solution :+1: